# Exploring Ebay Car Sales Data

This dataset contains 50000 data points of used cars from Ebay-Kleinanzeigen.
The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). 
The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.
The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which which year the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which which year the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings. 

In [1]:
import numpy as np
import pandas as pd
import operator

In [2]:
autos=pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset has 20 columns: 5 are integer types and the others are string. 5 columns have null values (vehicleType, gearbox, model, fuelType, notRepairedDamage), which are at most ~2% of the total values for that column, therefore we might be able to safely remove them (they are less than 20%) without compromising the validity of the dataset.

# Clean columns

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We clean the column names by changing them to more meaningful names and/or converting camel case to sneake case.

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

# Initial Data Exploration and Cleaning

We start by looking at descriptive statistics to check for anomalies and columns that might need cleaning

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

The initial observations are that `seller` and `offer_type` columns have only 2 unique values, meaning that almost all the vaues are the same. The `num_photos` seems to have all 0s and needs further exploration.

In [7]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

In [8]:
autos['seller'].unique()

array(['privat', 'gewerblich'], dtype=object)

`num_photos` has indeed all zeros, therefore we can drop this column, as well as `seller` and `offer_type` because they do not add valuable information.

In [9]:
autos = autos.drop(['num_photos', 'seller', 'offer_type'], axis=1)

The columns `price` and `odometer` are numeric values but are stored as strings. We can clean them from non-numeric character and convert them to integer.

In [10]:
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In order to convert them to integer, we have to clean the values from the non-numeric character, in this case the "$" and the ",".

In [11]:
autos['price']=autos['price'].str.replace('$','').str.replace(',','').astype(int)

Because we have removed the $ sign, we rename the column as `price_usd` so that we do not lose the information that the prices are in USD.

In [12]:
autos.rename({"price": "price_usd"}, axis=1, inplace=True)
autos['price_usd'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_usd, dtype: int64

In [13]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

Similarly to the price, we remove the non-numeric character, here are the "km" and ",".

In [14]:
autos['odometer']=autos['odometer'].str.replace('km','').str.replace(',','').astype(int)

We also rename `odometer` to `odometer_km` to keep the information that the figures represent km.

In [15]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

# Exploring the Odometer and Price Columns

We explore the `price` and `odometer_km` columns to check for possible outliers.

In [16]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The values are rounded to the nearest 1000 and most of the cars have more than 125000 km.

In [17]:
print("Unique values of price: ")
print(autos['price_usd'].unique().shape)
print("Descriptive statistics of price:")
print(round(autos['price_usd'].describe(),0))
print('Most frequent prices: ')
autos['price_usd'].value_counts().head()

Unique values of price: 
(2357,)
Descriptive statistics of price:
count       50000.0
mean         9840.0
std        481104.0
min             0.0
25%          1100.0
50%          2950.0
75%          7200.0
max      99999999.0
Name: price_usd, dtype: float64
Most frequent prices: 


0       1421
500      781
1500     734
2500     643
1000     639
Name: price_usd, dtype: int64

There are only 2357 unique prices, and there are 1421 prices of 0. Given the size of the dataset, this is a small percentage and we might remove the corresponding rows.
Below we display the prices in descending order.

In [18]:
autos['price_usd'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_usd, dtype: int64

Now we display the frequencies of the prices (as unique values).

In [19]:
autos['price_usd'].value_counts().sort_index().head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_usd, dtype: int64

Some cars are priced over 1 million and few have very low prices. The can leave the low prices, except the zero, assuming that they are starting bidding prices on eBay, and remove anything above 350000, because those prices seem unrealistic.

In [20]:
autos = autos[autos['price_usd'].between(1,350000)]
autos['price_usd'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_usd, dtype: float64

# Exploring the date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler

- `last_seen`: added by the crawler

- `ad_created`: from the website

- `registration_month`: from the website

- `registration_year`: from the website

The non-registration dates are stored as string.

In [21]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The values are formatted with the first 10 character repesenting the day and the remaining representing the hour. We can extract the first 10 character to isolate the day and disregard the hour. 

In [22]:
# normalize -> relative frequencies of the unique values
print(autos["date_crawled"].str[:10].unique().shape)
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

(34,)


2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [23]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_values()

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

The data have been crawled from the website everyday, regularly, since the percentage of post per day is approximately the same.

In [24]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The last three days include a higher proportion of `last_seen` of the ads, which is the day the ad was last seen online, probably because the ad was removed after it was sold. This proportion is very high (up to 10 times as the other days) and cannot be explained by a rise in selling alone.

In [25]:
print(autos["ad_created"].str[:10].unique().shape)
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

Ads were created over a period of about 10 months from 2015-06-11 to 2016-04-07

In [26]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There seems to be some incorrect values in the `registration_year`, the oldest value is in fact 1000 and the newest is 9999, both year are unrealistic. 

# Dealing with Incorrect Registration Year Data

We see what proportion of values are outside the range of year 1900 and 2016, which are the only years that look realistic.

In [27]:
outside_range = (~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]
print('Percentage of value years outside the range 1900-2016 %s%%' % round(outside_range*100, 2))

Percentage of value years outside the range 1900-2016 3.88%


The percentage of this values outside of the rage is very small (less than 4%) therefore we can safely remove those values.

In [28]:
autos=autos[autos['registration_year'].between(1900, 2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

After removing outlier years, we see that the most frequent registration years are from 1999 to 2007

# Exploring Price by Brand

We list the most frequent brand in the ads.

In [29]:
autos['brand'].value_counts(normalize = True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

We can aggregate the brand which have al least 5% of values.

In [30]:
brands=autos['brand'].value_counts(normalize = True)
selected_brands=brands[brands>.05].index # we select just the index names

We create a dictionary to store the mean price of each of the top selected brand.

In [31]:
brand_mean_prices = {}
for brand in selected_brands:
    brand_mean_prices[brand] = int(autos[autos["brand"] == brand]['price_usd'].mean())
    

In [32]:
sorted(brand_mean_prices.items(), key=operator.itemgetter(1), reverse=True)

[('audi', 9336),
 ('mercedes_benz', 8628),
 ('bmw', 8332),
 ('volkswagen', 5402),
 ('ford', 3749),
 ('opel', 2975)]

After aggregating the mean price for the most popular brands, we notice that: 
- `Audi`, `BMW` and `Mercedes Benz` are the most expensive cars 
- `Opel` and `Ford` are the least expensive 
- `Volkswagen` is in between

# Storing aggregate data in a dataframe

We can create a new dataframe, with the selected information that we extracted, such as the mean price. In this way we can compare only the columns that interest us.
First we create a series with the mean price, and then we include that series in the dataframe.

In [33]:
bmp_series = pd.Series(brand_mean_prices)
popular_brands = pd.DataFrame(bmp_series, columns=['mean_price'])

We calculate the mean mileage per brand and add it to our dataframe.

In [34]:
brand_mean_mileage = {}
for brand in selected_brands:
    brand_mean_mileage[brand] = int(autos[autos["brand"] == brand]['odometer_km'].mean())

In [35]:
popular_brands['mean_mileage']=pd.Series(brand_mean_mileage)

In [36]:
popular_brands

mean_price  mean_mileage
audi                 9336        129157
bmw                  8332        132572
ford                 3749        124266
mercedes_benz        8628        130788
opel                 2975        129310
volkswagen           5402        128707

We notice that the average mileage is consistent across the brands and does not have the same variation as the price.

# Further data cleaning

### Converting the date to integer

We can convert to integer also the dates, so that we can perform more analysis on them if needed, by treating them as numbers.
In order to do so, we extract the first 10 characters of the `date_crawled`, `ad_created` and `last_seen` (which represent the year-month-day), remove the non-numberic characters and then convert to integer.

In [37]:
def date_toInt(series,column_name):
    series[column_name]=series[column_name].str[:10].str.replace("-","").astype(int)    

In [38]:
date_toInt(autos,'date_crawled')
date_toInt(autos,'ad_created')
date_toInt(autos,'last_seen')

In [39]:
autos.describe().astype(int)

date_crawled  price_usd  registration_year  power_ps  odometer_km  \
count         46681      46681              46681     46681        46681   
mean       20160332       5977               2002       117       125586   
std              31       9177                  7       184        39852   
min        20160305          1               1910         0         5000   
25%        20160313       1250               1999        75       100000   
50%        20160321       3100               2003       109       150000   
75%        20160329       7500               2008       150       150000   
max        20160407     350000               2016     17700       150000   

       registration_month  ad_created  postal_code  last_seen  
count               46681       46681        46681      46681  
mean                    5    20160330        51097   20160369  
std                     3         110        25755         42  
min                     0    20150611         1067   20160305  
25%                     3    20160313        30827   20160323  
50%                     6    20160321        49828   20160404  
75%                     9    20160329        71732   20160406  
max                    12    20160407        99998   20160407

The descriptive statistics is now showing the oldest and newest dates for the columns we have modified.

### Translating German values to English

We are also going to check for categorical data that use German words and we will translate them to English.

In [40]:
autos.head(20)

date_crawled                                               name  \
0       20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   
1       20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2       20160326                         Volkswagen_Golf_1.6_United   
3       20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4       20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5       20160321  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6       20160320  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7       20160316                               Golf_IV_1.9_TDI_90PS   
8       20160322                                         Seat_Arosa   
9       20160316          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11      20160316                         Mercedes_A140_Motorschaden   
12      20160331  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13      20160323                                 Audi_A3_1.6_tuning   
14      20160323  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15      20160401        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16      20160316                                Opel_Vectra_B_Kombi   
17      20160329                          Volkswagen_Scirocco_2_G60   
18      20160326               Verkaufen_mein_bmw_e36_320_i_touring   
19      20160317        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20      20160305  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   

    price_usd  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0        5000  control          bus               2004    manuell       158   
1        8500  control    limousine               1997  automatik       286   
2        8990     test    limousine               2009    manuell       102   
3        4350  control   kleinwagen               2007  automatik        71   
4        1350     test        kombi               2003    manuell         0   
5        7900     test          bus               2006  automatik       150   
6         300     test    limousine               1995    manuell        90   
7        1990  control    limousine               1998    manuell        90   
8         250     test          NaN               2000    manuell         0   
9         590  control          bus               1997    manuell        90   
11        350  control          NaN               2000        NaN         0   
12       5299  control   kleinwagen               2010  automatik        71   
13       1350  control    limousine               1999    manuell       101   
14       3999     test   kleinwagen               2007    manuell        75   
15      18900     test        coupe               1982  automatik       203   
16        350     test        kombi               1999    manuell       101   
17       5500     test        coupe               1990    manuell       205   
18        300  control          bus               1995    manuell       150   
19       4150  control          suv               2004    manuell       124   
20       3500     test        kombi               2003    manuell       131   

       model  odometer_km  registration_month fuel_type           brand  \
0     andere       150000                   3       lpg         peugeot   
1        7er       150000                   6    benzin             bmw   
2       golf        70000                   7    benzin      volkswagen   
3     fortwo        70000                   6    benzin           smart   
4      focus       150000                   7    benzin            ford   
5    voyager       150000                   4    diesel        chrysler   
6       golf       150000                   8    benzin      volkswagen   
7       golf       150000                  12    diesel      volkswagen   
8      arosa       150000                  10       NaN            seat   
9     megane       150000                   7    benzin         renault   
11       NaN       150000  

It looks like `vehicle_type`, `gearbox`, `model`, `fuel_type`, `brand` and `unrepaired_damage` have some German words, we are going to check that in details which words need to be changed and replace them with English translations.

In [41]:
autos['vehicle_type'].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [42]:
autos['gearbox'].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [43]:
autos['model'].unique()

array(['andere', '7er', 'golf', 'fortwo', 'focus', 'voyager', 'arosa',
       'megane', nan, 'a3', 'clio', 'vectra', 'scirocco', '3er', 'a4',
       '911', 'cooper', '5er', 'polo', 'e_klasse', '2_reihe', 'c_klasse',
       'corsa', 'mondeo', 'altea', 'a1', 'twingo', 'a_klasse', 'cl',
       '3_reihe', 's_klasse', 'sandero', 'passat', 'primera', 'wrangler',
       'a6', 'transporter', 'astra', 'v40', 'ibiza', 'micra', '1er',
       'yaris', 'colt', '6_reihe', '5_reihe', 'corolla', 'ka', 'tigra',
       'punto', 'vito', 'cordoba', 'galaxy', '100', 'octavia', 'm_klasse',
       'lupo', 'fiesta', 'superb', 'meriva', 'c_max', 'laguna', 'touran',
       '1_reihe', 'm_reihe', 'touareg', 'seicento', 'avensis', 'vivaro',
       'x_reihe', 'ducato', 'carnival', 'boxster', 'signum', 'sharan',
       'zafira', 'rav', 'a5', 'beetle', 'c_reihe', 'phaeton', 'i_reihe',
       'sl', 'insignia', 'up', 'civic', '80', 'mx_reihe', 'omega',
       'sorento', 'z_reihe', 'berlingo', 'clk', 's_max', 'kalos',
 

In [44]:
autos['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [45]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [46]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In [47]:
columns = ['vehicle_type', 'gearbox', 'fuel_type', 'model', 'unrepaired_damage', 'brand']
de_to_en = {"kleinwagen": "super mini", "kombi":"station wagon", "cabrio": "convertible", "limousine": "sedan", "andere": "other", "manuell": "manual", "automatik": "automatic", "benzin":"gas", "elektro": "electric", "sonstige_autos": "other", "nein": "no", "ja": "yes"}
autos[columns]=autos[columns].replace(de_to_en)

In [48]:
autos.head(20)

date_crawled                                               name  \
0       20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   
1       20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2       20160326                         Volkswagen_Golf_1.6_United   
3       20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4       20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5       20160321  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6       20160320  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7       20160316                               Golf_IV_1.9_TDI_90PS   
8       20160322                                         Seat_Arosa   
9       20160316          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11      20160316                         Mercedes_A140_Motorschaden   
12      20160331  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13      20160323                                 Audi_A3_1.6_tuning   
14      20160323  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15      20160401        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16      20160316                                Opel_Vectra_B_Kombi   
17      20160329                          Volkswagen_Scirocco_2_G60   
18      20160326               Verkaufen_mein_bmw_e36_320_i_touring   
19      20160317        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20      20160305  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   

    price_usd  ab_test   vehicle_type  registration_year    gearbox  power_ps  \
0        5000  control            bus               2004     manual       158   
1        8500  control          sedan               1997  automatic       286   
2        8990     test          sedan               2009     manual       102   
3        4350  control     super mini               2007  automatic        71   
4        1350     test  station wagon               2003     manual         0   
5        7900     test            bus               2006  automatic       150   
6         300     test          sedan               1995     manual        90   
7        1990  control          sedan               1998     manual        90   
8         250     test            NaN               2000     manual         0   
9         590  control            bus               1997     manual        90   
11        350  control            NaN               2000        NaN         0   
12       5299  control     super mini               2010  automatic        71   
13       1350  control          sedan               1999     manual       101   
14       3999     test     super mini               2007     manual        75   
15      18900     test          coupe               1982  automatic       203   
16        350     test  station wagon               1999     manual       101   
17       5500     test          coupe               1990     manual       205   
18        300  control            bus               1995     manual       150   
19       4150  control            suv               2004     manual       124   
20       3500     test  station wagon               2003     manual       131   

       model  odometer_km  registration_month fuel_type          brand  \
0      other       150000                   3       lpg        peugeot   
1        7er       150000                   6       gas            bmw   
2       golf        70000                   7       gas     volkswagen   
3     fortwo        70000                   6       gas          smart   
4      focus       150000                   7       gas           ford   
5    voyager       150000                   4    diesel       chrysler   
6       golf       150000                   8       gas     volkswagen   
7       golf       150000                  12    diesel     volkswagen   
8      arosa       150000                  10       NaN           seat   
9     megane       150000                   7       gas        renault

# Further analysis

### Finding the most common brand/model combinations

Previously we have looked at the most common brands, we can do a further analysis on this and add the model to our search, therefore finding the most common brand/model combinations.

In [49]:
brand_models = autos.loc[autos['brand'].isin(selected_brands)].groupby(['brand', 'model']).size().sort_values(ascending=False)
brand_models.head(10)

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1592
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
dtype: int64

Three models of Volkswagen are in the top 10, unsurprisingly since we knew that Volkswagen is the most popular brand in this dataset. The most popular model is Volkswagen Golf. Bmw, Opel and Mercedes Benz, have each two models in the top ten.

We build a table to see the top 3 models for each of the top advertised brand.

In [50]:
top3_brand_models = (brand_models.reset_index() #creates a basic table
                          .sort_values(['brand',0], ascending=[True,False]) #sorting by brand and then by count
                          .set_index(['brand','model']) #this will make showing the brand only once
                          .groupby('brand').head(3) # restricting to the top 3
                    .rename(columns={0: 'count'})) 
top3_brand_models

count
brand         model          
audi          a4         1231
              a3          825
              a6          797
bmw           3er        2615
              5er        1132
              1er         521
ford          focus       762
              fiesta      722
              mondeo      479
mercedes_benz c_klasse   1136
              e_klasse    958
              a_klasse    539
opel          corsa      1592
              astra      1348
              vectra      544
volkswagen    golf       3707
              polo       1609
              passat     1349

### How much cheaper are cars with damage than their non-damaged counterparts?

In [51]:
price_damaged = autos.groupby(['unrepaired_damage'])['price_usd', 'brand'].mean().astype(int).reset_index().rename(columns={'price_usd': 'avg_price_usd'})
price_damaged

unrepaired_damage  avg_price_usd
0                no           7164
1               yes           2241

Looking at the average prices of cars with unrepaired damages, it is clear that those cars are a lot cheaper than those who do not have unrepaired damages (more than 3 times cheaper on average). 

# Conclusions

In this project we have performed explorative analysis, after cleaning the data, of a subset of used cars ads from a German advertising website.
The main findings are:
- The brands which are most advertised are `Volkswagen`, `Bmw`, `Opel`, `Mercedes_benz`, `Audi` and `Ford`. Most of which are German brands, and being this dataset from a German website;
- `Volkswagen` is also on average the a middle price car;
- The most popular car overall is `Volkswagen Golf`;
- Cars with unrepaired damages are up to 66% cheaper than those which do not have unrepaired damages;

Further analysis might include: 
- Exploring the correlation between the age of the car and it's price; 
- Checking which brand sells faster (e.g. by looking at date of ad creating and last seen date);
- Finding correlations between prices and locations (according to postcode), in order to see if cars are more expensive in certain areas.